README

rfkill applet
version 0.8

Access to the rfkill switches on the system

Warning: I never programmed python, nor pygtk, never any applet, never
dbus/hal, so please do expect this applet to make your computer disappear
in flames.

Code sources and thanks:
- gfreqlet.sf.net for pygtk applets
- dbus/hal interface by trial and error with loads of help from the
  plasmoid-applet-killswitch of Eva Brucherseifer
  (see http://www.basyskom.org/~eva/log_installation_vaio_z21vnx.html)
- rfkill access mode from the source code of rfkill user space program
  and thanks to hints by Johannes Berg <johannes AT sipsolutions DOT net>


Installation:
- you have to install Python interface to GNOME Applet installed
  On Debian/Ubuntu the package name is python-gnomeapplet
- put rfkill-applet.py into /usr/bin/rfkill-applet  (no .py extension!)
  and make it executable (chmod a+x /usr/bin/rfkill-applet)
- put rfkill-applet.server into /usr/lib/bonobo/servers/ and make it
  executable (chmod a+x /usr/lib/bonobo/servers/rfkill-applet.server)
- put rfkill-applet.png into /usr/share/pixmaps
- put rfkill-applet-hardoff.png into /usr/share/pixmaps
- put rfkill-applet.config into 
     either /etc/rfkill-applet.config
     or     $HOME/.rfkill-applet.config
- restart gnome session
- add the applet

Sys file access:
Since version 0.8 rfkill-applet can be used to turn on/off "settings"
in the /sys file hierarchy, of in fact, everywhere. See examples
that make sense for the sony laptop in the example configuration file.
If you need write access as user you have to specify some udev rules
that allow write access to these files.

By default rfkill-applet will first try to use /dev/rfkill if available
and read/writeable. This access method is the preferred one as it is
less resource intensive (no need to wake up every 3 secs). This interface
to the rfkill subsystem was introduced in the kernel in 2.6.31-rc series.
In case /dev/rfkill is not available or not useable rfkill-applet falls
back to HAL/DBUS access.

To configure /dev/rfkill access I recommend:
- adding a group rfkill
- adding the users that should be allowed to change rfkills to this group
- adding a line to one of the config files for udev (or creating a new one),
  eg /etc/udev/rules.d/99-local.rules with one line:
  	KERNEL=="rfkill", NAME="rfkill", OWNER="root", GROUP="rfkill", MODE="660"


HAL/DBUS is normally already configured so that each user can access
the rfkill stuff.


Bugs: many, help needed, suggestions welcome!

Copyright (C) 2009, 2010, 2011 Norbert Preining <preining@logic.at>
Licensed under GPLv3



CHANGES

2011-08 onward

  * add a few remarks to the README file about sys files
  * make sure that the applet does not crash on a not
    existing sys file

2011-07-26

  * release 0.8
  * allow to add switches linked to files on the filesystem, like
    /sys/devices/platform/sony_laptop/.... See example config file
    for how to use

2010-08-01

  * release 0.7
  * allow glob patterns in <NAME> in the <NAME>.ignore=1 config file.
    this allows ignoring all phy at once:
      phy*.ignore=1

2009-11-06

  * fix udev rule to give correct permissions
  * add rfkill-sys.py code from Andreas Boehler, it will later on 
    integrated into the normal code

2009-08-10

  * release 0.6
  * refactor dbus and /dev/rfkill access in separate objects
  * get rid of threading, use gobject.io_add_watch on the fd
    that means also that the rfkillclient.py is now outdated
    and should be removed (thanks to Johannes Berg)

2009-08-03

  * release 0.5
  * include support for using /dev/rfkill instead of HAL/DBUS if the
    user has r/w permissions for it.
  * fix icons for quit and about

2009-06-23

  * release 0.4

2009-06-20

  * change hardware switch detection logic: now there is no special hardware
    switch property, but if *one* value has the hardoff value it is assumed
    that the hardware switch is turned on, otherwise turned off.
    Adjust the config file removing the hardswitch directive, and ignore
    the 5100AGN switch
    WARNING: You have to upate your config files!

2009-06-20

  * release 0.3
  * check every 3 secs for updates
    that could be done better maybe by changing the event loop using
    dbus/glib and registering a callback with dbus when the rfkill 
    hardware switch state has changed
  * allow rfkill.name=foobar in config file, so to rename skitches
    adjust config file to strip the sony- prefix from the names

2009-06-20

  * release 0.2

2009-06-19

  * change icon if hardware switch is off
  * do not pop up the menu if the hard switch is activated
  * add a config file, and reading from it, either from 
    /etc/rfkill-applet.config or $HOME/.rfkill-applet.config

2009-06-18

   * pop up a message window telling that the prefs dialog is not implemented
     fix location of pixmap in the panel itself, need full path

2009-06-17

   * release 0.1: first release



rfkill-sys.py
#!/usr/bin/env python

# rfkill-sys.py
# (c) 2009 Andreas Boehler, andy (dot) boehler (at) gmx (dot) at
# heavily based on rfkill-applet.py by Norbert Preining, 
# 
# This program is OpenSource, licensed under GPLv3 or any later version
#
# Changelog:
#
# v0.1, 2009/11/02
# Initial version, in fact rfkill-applet with changed GUI

# Bugs/Todo:
# In /dev/rfkill-mode, the initial status is uknown!

import pygtk
import gtk
import gobject
import os
import struct
import dbus
import time

event_format = "IBBBB"

version='0.1'

RFKILL_OP_ADD = 0
RFKILL_OP_DEL = 1
RFKILL_OP_CHANGE = 2
RFKILL_OP_CHANGE_ALL = 3

class RfkillAccess():
    AccessO = None

    def __init__ (self, applet, ignore):
        self.applet = applet
        self.rfkill_names = dict()
        self.rfkill_hardstates = dict()
        self.rfkill_softstates = dict()
        try:
            f = os.open("/dev/rfkill", os.O_RDWR)
            os.close(f)
            self.AccessO = RfkillAccessDevRfkill(applet, ignore)
            print "Using /dev/rfkill access mode"

        except:
            self.AccessO = RfkillAccessDbus(applet, ignore)
            print "Using Hal/DBUS access mode"

    def get_state (self, idx):
        return self.AccessO.get_state(idx)

    def toggle_softstate (self, idx):
        return self.AccessO.toggle_softstate (idx)

    def get_rfkillall (self):
        return self.AccessO.get_rfkillall()


class RfkillAccessDbus():
    def __init__ (self, applet, ignore):
        self.applet = applet
        self.ignored = dict()
        for k,v in ignore.iteritems():
            if v:
                self.ignored[k] = 1
        self.bus = dbus.SystemBus()
        self.hal_obj = self.bus.get_object("org.freedesktop.Hal", "/org/freedesktop/Hal/Manager")
        self.hal = dbus.Interface(self.hal_obj, "org.freedesktop.Hal.Manager")
        self.rfkill_names = dict()
        self.rfkill_hardstate = dict()
        self.rfkill_softstate = dict()
        # set up timeout for restarting
        gobject.timeout_add(3000, self.periodic_check)
        
    def periodic_check (self):
        self.parent_set_hard_switch()
        return True

    def parent_set_hard_switch (self):
        # will only be called in HAL/DBUS mode
        # we have to check for the actual hard switches
        saved_hard = self.rfkill_hardstate
        allrf = self.get_rfkillall()
        is_hard_off = False
        for idx, name in self.get_rfkillall().iteritems():
            if not(name in self.ignored):
                newhard, newsoft = self.get_state(idx)
                if newhard:
                    is_hard_off = True
        self.applet.set_hard_switch(is_hard_off)
      
    def get_rfkillall (self):
        self.rfkill_names = dict()
        self.rfkill_hardstate = dict()
        self.rfkill_softstate = dict()
        self.rfkill_devs = dict()
        for udi in self.hal.FindDeviceByCapability ("killswitch"):
            dev_obj = self.bus.get_object('org.freedesktop.Hal', udi)
            dev = dbus.Interface(dev_obj, 'org.freedesktop.Hal.Device')
            if (dev.GetProperty('killswitch.type') != "unknown"):
                name = str(dev.GetProperty ('killswitch.name'))
                if (name in self.ignored):
                    continue
                self.rfkill_names[udi] = name
                val = int(dev.GetProperty ('killswitch.state'))
                self.rfkill_devs[udi] = dev_obj
                if (val == 1):
                    self.rfkill_hardstate[udi] = 0
                    self.rfkill_softstate[udi] = 0
                elif (val == 2):
                    self.rfkill_softstate[udi] = 0
                    self.rfkill_hardstate[udi] = 1
                else:
                    self.rfkill_softstate[udi] = 1
                    self.rfkill_hardstate[udi] = 0
        return (self.rfkill_names)
     
    def get_state (self, idx):
        if self.rfkill_names[idx]:
            return(self.rfkill_hardstate[idx], self.rfkill_softstate[idx])
        else:
            return None

    def toggle_softstate (self, idx):
        dev = dbus.Interface(self.rfkill_devs[idx], 'org.freedesktop.Hal.Device.KillSwitch')
        # the value is already inverted, so no need for not
        dev.SetPower(self.rfkill_softstate[idx])

class RfkillAccessDevRfkill():
    def __init__(self, applet, ignore):
        self.ignored = dict()
        self.applet = applet
        for k,v in ignore.iteritems():
            if v:
                self.ignored[k] = 1
        self.rfkillfd = os.open("/dev/rfkill", os.O_RDONLY)
        gobject.io_add_watch(self.rfkillfd, gobject.IO_IN, self.callback_event)
        self.rfkill_names = dict()
        self.rfkill_hardstate = dict()
        self.rfkill_softstate = dict()

    def callback_event (self, fd, condition):
        buf = os.read(self.rfkillfd, 8)
        if (len(buf) != 8):
            print "cannot read full event from fd"
        else:
            (idx, type, op, soft, hard) = struct.unpack(event_format, buf)
        if op == RFKILL_OP_DEL:
            if (idx in self.rfkill_names):
                del self.rfkill_names[idx]
            return True
        if op == RFKILL_OP_ADD:
            # name should be set from /sys/class/rfkill/rfkill%N/name
            f = open("/sys/class/rfkill/rfkill" + str(idx) + "/name")
            name = f.readline().rstrip()
            f.close()
            if not(name in self.ignored):
                self.rfkill_names[idx] = name

        if (idx in self.rfkill_names):
            # otherwise the rfkill switch is ignored

            # we want to check whether a hard switch was toggled
            # this is done by checking whether the property hard has
            # changed
            if (idx in self.rfkill_hardstate):
                if (hard != self.rfkill_hardstate[idx]):
                    # inform parent that we have a hard switch toggle
                    self.applet.set_hard_switch(hard)
            self.rfkill_hardstate[idx] = hard
            self.rfkill_softstate[idx] = soft

        return True

    def parent_set_hard_switch (self):
        return None
      
    def get_rfkillall (self):
        return (self.rfkill_names)
 
    def get_state (self, idx):
        if self.rfkill_names[idx]:
            return(self.rfkill_hardstate[idx], self.rfkill_softstate[idx])
        else:
            return None

    def toggle_softstate (self, idx):
        buf = struct.pack(event_format, idx, 0, RFKILL_OP_CHANGE, 
                      not(self.rfkill_softstate[idx]), 0)
        writefd = os.open("/dev/rfkill", os.O_RDWR)
        if (os.write(writefd, buf) < 8):
            print "Cannot write to rfkill the full event type"
        os.close(writefd)


class TrayMenu(object):

    def __init__(self):
        self.config_ignore = {}
        self.config_names = {}
        self.config_ignore = {}
        self.configfile = os.environ.get('HOME') + '/.rfkill-applet.config'
        self.read_config('/etc/rfkill-applet.config')
        self.read_config(self.configfile) 
        self.Access0 = RfkillAccess(self, self.config_ignore)
        self.update_all()
        
    def read_config(self, file):
        try:
            cfg = open(file, 'r')
            cfgdata = cfg.readlines()
            cfg.close()
        except:
            return

        for line in cfgdata:
            if (line.lstrip().startswith('#')):
                continue
            if line.strip() == '':
                continue
            key, val = line.strip().split('=',1)
            if val != '':
                rf,prop = key.split('.',1)
                if prop == 'onvalue':
                    self.onvalue[rf] = int(val)
                elif prop == 'offvalue':
                    self.offvalue[rf] = int(val)
                elif prop == 'ignore':
                    self.config_ignore[rf] = val
                elif prop == 'name':
                    self.config_names[rf] = val
                else:
                    print "Unkown key in config file: " + line

    def update_all(self):
        self.rfkills_hard = []
        self.rfkills_soft = []
        self.rfkills_name = []
        self.rfkills_idx = []
        self.rfkills_showname = []
        self.hardswitchedoff = False
        for idx, name in self.Access0.get_rfkillall().iteritems():
            hard, soft = self.Access0.get_state(idx)
            if hard:
                self.hardswitchedoff = True
                self.set_hard_switch(True)
            self.rfkills_hard.append(hard)
            self.rfkills_soft.append(soft)
            self.rfkills_name.append(name)
            self.rfkills_idx.append(idx)
            if (name in self.config_names):
               self.rfkills_showname.append(self.config_names[name])
            else:
                self.rfkills_showname.append(name)        
        
    def quit_activate(self, widget):
        gtk.main_quit()
        
    def set_hard_switch (self, newstate):
        self.hardswitchedoff = newstate
        if not newstate:
            icon.set_from_icon_name('rfkill-applet')
            icon.set_tooltip('Change RfKill Status')
        else:
            icon.set_from_icon_name('rfkill-applet-hardoff')  
            icon.set_tooltip('Hardware RfKill Switch is active!')
        
    def show_menu(self, widget, button, time):
        menu = gtk.Menu()
        self.update_all()
        if not self.hardswitchedoff:
            for idx,showname in enumerate(self.rfkills_showname):
                item = gtk.CheckMenuItem(label=showname)
                item.set_active(not(self.rfkills_soft[idx]))
                item.connect('activate', self.toggle_rfkill, idx)
                menu.append(item)
        item = gtk.ImageMenuItem(stock_id=gtk.STOCK_QUIT)
        item.connect('activate', self.quit_activate)
        menu.append(item)                
        menu.show_all()
        menu.popup(None, None, gtk.status_icon_position_menu, button, time, icon)
            
    def toggle_rfkill(self, widget, idx):
        self.Access0.toggle_softstate(self.rfkills_idx[idx])


icon = gtk.status_icon_new_from_icon_name('rfkill-applet')
icon.set_tooltip('RfKill')
menu = TrayMenu()
icon.connect('popup-menu', menu.show_menu)
#icon.connect('activate', menu.show_menu) # Does not work, because not enough arguments are emitted

gtk.main()


rfkillclient.py
#!/usr/bin/env python
#
# rfkill-client
# (C) 2009 Norbert Preining
# Licensed under GPLv3 or any version higher
#
# This is the thread that reads the rfkill device and caches the current
# status
#

import sys
import os
import threading
import select
import struct
import dbus
import time

event_format = "IBBBB"

RFKILL_OP_ADD = 0
RFKILL_OP_DEL = 1
RFKILL_OP_CHANGE = 2
RFKILL_OP_CHANGE_ALL = 3

class RfkillClient(threading.Thread):
  rfkillfd = None
  PollObj = None
  die = False

  def __init__(self, applet, ignore):

    super(RfkillClient, self).__init__()
    self.applet = applet
    self.devrfkill_works = False
    self.ignored = dict()
    for k,v in ignore.iteritems():
      if v:
        self.ignored[k] = 1

    try:
      f = os.open("/dev/rfkill", os.O_RDWR)
      self.devrfkill_works = True
      os.close(f)
      self.rfkillfd = os.open("/dev/rfkill", os.O_RDONLY)
      print "Using /dev/rfkill access mode"

    except:
      self.devrfkill_works = False
      print "Using Hal/DBUS access mode"

    if self.devrfkill_works:
      self.PollObj = select.poll()
      self.PollObj.register(self.rfkillfd, select.POLLIN | select.POLLHUP)
      self.die = False
    else:
      # why does this not work here? Why is it impossible to load
      # on demand?
      # import dbus
      self.bus = dbus.SystemBus()
      self.hal_obj = self.bus.get_object("org.freedesktop.Hal", "/org/freedesktop/Hal/Manager")
      self.hal = dbus.Interface(self.hal_obj, "org.freedesktop.Hal.Manager")


    self.rfkill_names = dict()
    self.rfkill_hardstate = dict()
    self.rfkill_softstate = dict()

  def run (self):
    
    if not(self.devrfkill_works):
      while self.die == False:
        # HAL/DBUS we wake up periodically and inform parent on the state
        time.sleep(3)
        self.parent_set_hard_switch()
    else:
      while self.die == False:
        n = self.PollObj.poll()
        if (n[0]):
          if (n[0][0] == self.rfkillfd):
            t = n[0][1]
            if (t == select.POLLIN):
              buf = os.read(self.rfkillfd, 8)
              if (len(buf) != 8):
                print "cannot read full event from fd"
              else:
                (idx, type, op, soft, hard) = struct.unpack(event_format, buf)
                if op == RFKILL_OP_DEL:
                  if (idx in self.rfkill_names):
                    del self.rfkill_names[idx]
                  continue
                  next
                if op == RFKILL_OP_ADD:
                  # name should be set from /sys/class/rfkill/rfkill%N/name
                  f = open("/sys/class/rfkill/rfkill" + str(idx) + "/name")
                  name = f.readline().rstrip()
                  f.close()
                  if not(name in self.ignored):
                    self.rfkill_names[idx] = name

                if (idx in self.rfkill_names):
                  # otherwise the rfkill switch is ignored

                  # we want to check whether a hard switch was toggled
                  # this is done by checking whether the property hard has
                  # changed
                  if (idx in self.rfkill_hardstate):
                    if (hard != self.rfkill_hardstate[idx]):
                      # inform parent that we have a hard switch toggle
                      self.applet.set_hard_switch(hard)
                  self.rfkill_hardstate[idx] = hard
                  self.rfkill_softstate[idx] = soft

            #elif (t == select.POLLHUP):
              # no idea what we can do here, or should do here
              # print "pollhup event"
            else:
              print "unknown event"

  def parent_set_hard_switch (self):
    # will only be called in HAL/DBUS mode
    # we have to check for the actual hard switches
    saved_hard = self.rfkill_hardstate
    allrf = self.get_rfkillall
    is_hard_off = False
    for idx, name in self.get_rfkillall().iteritems():
      if not(name in self.ignored):
        newhard, newsoft = self.get_state(idx)
        if newhard:
          is_hard_off = True
    self.applet.set_hard_switch(is_hard_off)
      

  def get_rfkillall (self):
    if not(self.devrfkill_works):
      # the HAL case, so nothing initialized already
      # we need to wade through the HAL stuff
      self.rfkill_names = dict()
      self.rfkill_hardstate = dict()
      self.rfkill_softstate = dict()
      self.rfkill_devs = dict()
      for udi in self.hal.FindDeviceByCapability ("killswitch"):
        dev_obj = self.bus.get_object('org.freedesktop.Hal', udi)
        dev = dbus.Interface(dev_obj, 'org.freedesktop.Hal.Device')
        if (dev.GetProperty('killswitch.type') != "unknown"):
          name = str(dev.GetProperty ('killswitch.name'))
          if (name in self.ignored):
            continue
          self.rfkill_names[udi] = name
          val = int(dev.GetProperty ('killswitch.state'))
          self.rfkill_devs[udi] = dev_obj
          if (val == 1):
            self.rfkill_hardstate[udi] = 0
            self.rfkill_softstate[udi] = 0
          elif (val == 2):
            self.rfkill_softstate[udi] = 0
            self.rfkill_hardstate[udi] = 1
          else:
            self.rfkill_softstate[udi] = 1
            self.rfkill_hardstate[udi] = 0
    return (self.rfkill_names)
 
  def get_state (self, idx):
    if self.rfkill_names[idx]:
      return(self.rfkill_hardstate[idx], self.rfkill_softstate[idx])
    else:
      return None

  def toggle_softstate (self, idx):
    if (self.devrfkill_works):
      buf = struct.pack(event_format, idx, 0, RFKILL_OP_CHANGE, 
                        not(self.rfkill_softstate[idx]), 0)
      writefd = os.open("/dev/rfkill", os.O_RDWR)
      if (os.write(writefd, buf) < 8):
        print "Cannot write to rfkill the full event type"
      os.close(writefd)
    else:
      dev = dbus.Interface(self.rfkill_devs[idx], 'org.freedesktop.Hal.Device.KillSwitch')
      # the value is already inverted, so no need for not
      dev.SetPower(self.rfkill_softstate[idx])

 
  def kill(self):
    self.die = True


# vim:set tabstop=2 expandtab: #


rfkill-applet.py
#!/usr/bin/env python
#
# rfkill-applet
# (C) 2009, 2010, 2011 Norbert Preining
# Licensed under GPLv3 or any version higher
#

import sys
import os
import pygtk
pygtk.require('2.0')
import gtk
import gio
import gnomeapplet
import gobject
import struct
import dbus
import fnmatch

version = '0.9pre'


event_format = "IBBBB"

RFKILL_OP_ADD = 0
RFKILL_OP_DEL = 1
RFKILL_OP_CHANGE = 2
RFKILL_OP_CHANGE_ALL = 3

def is_ignore_with_wildcards (what, ilist):
  for dname in ilist:
    if fnmatch.fnmatch(what, dname):
      return True
  return False

class RfkillAccess():
  AccessO = None

  def __init__ (self, applet, ignore):
    self.applet = applet
    self.rfkill_names = dict()
    self.rfkill_hardstates = dict()
    self.rfkill_softstates = dict()
    try:
      f = os.open("/dev/rfkill", os.O_RDWR)
      os.close(f)
      self.AccessO = RfkillAccessDevRfkill(applet, ignore)
      print "Using /dev/rfkill access mode"

    except:
      self.AccessO = RfkillAccessDbus(applet, ignore)
      print "Using Hal/DBUS access mode"

  def get_state (self, idx):
    return self.AccessO.get_state(idx)

  def toggle_softstate (self, idx):
    return self.AccessO.toggle_softstate (idx)

  def get_rfkillall (self):
    return self.AccessO.get_rfkillall()


class RfkillAccessDbus():
  def __init__ (self, applet, ignore):
    self.applet = applet
    self.ignored = dict()
    for k,v in ignore.iteritems():
      if v:
        self.ignored[k] = 1
    self.bus = dbus.SystemBus()
    self.hal_obj = self.bus.get_object("org.freedesktop.Hal", "/org/freedesktop/Hal/Manager")
    self.hal = dbus.Interface(self.hal_obj, "org.freedesktop.Hal.Manager")
    self.rfkill_names = dict()
    self.rfkill_hardstate = dict()
    self.rfkill_softstate = dict()
    # set up timeout for restarting
    gobject.timeout_add(3000, self.periodic_check)

  def periodic_check (self):
    self.parent_set_hard_switch()
    return True

  def parent_set_hard_switch (self):
    # will only be called in HAL/DBUS mode
    # we have to check for the actual hard switches
    saved_hard = self.rfkill_hardstate
    allrf = self.get_rfkillall()
    is_hard_off = False
    for idx, name in self.get_rfkillall().iteritems():
      if not(is_ignore_with_wildcards(name, self.ignored)):
        newhard, newsoft = self.get_state(idx)
        if newhard:
          is_hard_off = True
    self.applet.set_hard_switch(is_hard_off)
      
  def get_rfkillall (self):
    self.rfkill_names = dict()
    self.rfkill_hardstate = dict()
    self.rfkill_softstate = dict()
    self.rfkill_devs = dict()
    for udi in self.hal.FindDeviceByCapability ("killswitch"):
      dev_obj = self.bus.get_object('org.freedesktop.Hal', udi)
      dev = dbus.Interface(dev_obj, 'org.freedesktop.Hal.Device')
      if (dev.GetProperty('killswitch.type') != "unknown"):
        name = str(dev.GetProperty ('killswitch.name'))
        if is_ignore_with_wildcards(name, self.ignored):
          continue
        self.rfkill_names[udi] = name
        val = int(dev.GetProperty ('killswitch.state'))
        self.rfkill_devs[udi] = dev_obj
        if (val == 1):
          self.rfkill_hardstate[udi] = 0
          self.rfkill_softstate[udi] = 0
        elif (val == 2):
          self.rfkill_softstate[udi] = 0
          self.rfkill_hardstate[udi] = 1
        else:
          self.rfkill_softstate[udi] = 1
          self.rfkill_hardstate[udi] = 0
    return (self.rfkill_names)
 
  def get_state (self, idx):
    if self.rfkill_names[idx]:
      return(self.rfkill_hardstate[idx], self.rfkill_softstate[idx])
    else:
      return None

  def toggle_softstate (self, idx):
    dev = dbus.Interface(self.rfkill_devs[idx], 'org.freedesktop.Hal.Device.KillSwitch')
    # the value is already inverted, so no need for not
    dev.SetPower(self.rfkill_softstate[idx])



class RfkillAccessDevRfkill():
  def __init__(self, applet, ignore):
    self.ignored = dict()
    self.applet = applet
    for k,v in ignore.iteritems():
      if v:
        self.ignored[k] = 1
    self.rfkillfd = os.open("/dev/rfkill", os.O_RDONLY)
    gobject.io_add_watch(self.rfkillfd, gobject.IO_IN, self.callback_event)
    self.rfkill_names = dict()
    self.rfkill_hardstate = dict()
    self.rfkill_softstate = dict()

  def callback_event (self, fd, condition):
    buf = os.read(self.rfkillfd, 8)
    if (len(buf) != 8):
      print "cannot read full event from fd"
    else:
      (idx, type, op, soft, hard) = struct.unpack(event_format, buf)
      if op == RFKILL_OP_DEL:
        if (idx in self.rfkill_names):
          del self.rfkill_names[idx]
        return True
      if op == RFKILL_OP_ADD:
        # name should be set from /sys/class/rfkill/rfkill%N/name
        f = open("/sys/class/rfkill/rfkill" + str(idx) + "/name")
        name = f.readline().rstrip()
        f.close()
        if not(is_ignore_with_wildcards(name, self.ignored)):
          self.rfkill_names[idx] = name

      if (idx in self.rfkill_names):
        # otherwise the rfkill switch is ignored

        # we want to check whether a hard switch was toggled
        # this is done by checking whether the property hard has
        # changed
        if (idx in self.rfkill_hardstate):
          if (hard != self.rfkill_hardstate[idx]):
            # inform parent that we have a hard switch toggle
            self.applet.set_hard_switch(hard)
        self.rfkill_hardstate[idx] = hard
        self.rfkill_softstate[idx] = soft

    return True

  def parent_set_hard_switch (self):
    return None
      
  def get_rfkillall (self):
    return (self.rfkill_names)
 
  def get_state (self, idx):
    if self.rfkill_names[idx]:
      return(self.rfkill_hardstate[idx], self.rfkill_softstate[idx])
    else:
      return None

  def toggle_softstate (self, idx):
    buf = struct.pack(event_format, idx, 0, RFKILL_OP_CHANGE, 
                      not(self.rfkill_softstate[idx]), 0)
    writefd = os.open("/dev/rfkill", os.O_RDWR)
    if (os.write(writefd, buf) < 8):
      print "Cannot write to rfkill the full event type"
    os.close(writefd)
 

class SysSwitch:
  def __init__(self, applet, name, filename, onval):
    self.applet = applet
    self.fn = filename
    self.name = name
    self.onvalue = onval
    self.giof = gio.File(path=filename)
    self.monitor = self.giof.monitor_file(gio.FILE_MONITOR_NONE, None)
    self.monitor.connect("changed", self.callback_event)
    self.state = int(self.get_state())

  def get_sysfs_value(self):
    try:
      file_descriptor = open(self.fn)
    except:
      return ""
    value = file_descriptor.read()
    file_descriptor.close()
    return value

  def set_sysfs_value(self, value):
    try:
      file_descriptor = open(self.fn, 'w')
    except:
      print "Cannot opfen " + self.fn + " for writing!"
      return False
    file_descriptor.write("%d" % int(value))
    file_descriptor.close()
    return True

  def callback_event (self, monitor, file1, file2, evt_type):
    self.state = int(self.get_state())
    return self.state

  def get_state (self):
    return self.get_sysfs_value()

  def toggle_softstate (self, idx):
    if (self.state == 0):
      newstate = self.onvalue
    else:
      newstate = 0
    if (self.set_sysfs_value(newstate)):
      self.state = newstate
      return True
    else:
      print "Cannot write to " + self.fn
      return False

class Rfkill:
  def __init__(self, applet, iid):

    self.configfile = os.environ.get('HOME') + '/.rfkill-applet.config'
    self.image = '/usr/share/pixmaps/rfkill-applet.png'
    self.imagehardoff = '/usr/share/pixmaps/rfkill-applet-hardoff.png'
    self.icon_hardon = gtk.Image()
    self.icon_hardon.set_from_file(self.image)
    self.icon_hardoff = gtk.Image()
    self.icon_hardoff.set_from_file(self.imagehardoff)

    self.config_names = {}
    self.config_files = {}
    self.config_types = {}
    self.config_ignore = {}
    self.onvalue = {}

    self.rfkills_hard = []
    self.rfkills_soft = []
    self.rfkills_name = []
    self.rfkills_showname = []
    self.hardswitchedoff = False

    self.sys_kills = []

    self.panel_size = 24

    self.applet = applet

    # read first the global config file, then a local one
    self.read_config('/etc/rfkill-applet.config')
    self.read_config(self.configfile)

    self.ebmain = gtk.EventBox()
    self.icon = gtk.Image()
    self.ebmain.add(self.icon)
    self.applet.add(self.ebmain)
    self.ebmain.connect("button-press-event", self.click_menu)
    applet.connect("destroy", self.cleanup)

    self.menuxml="""
    <popup name="button3">
    <menuitem name="Item 1" verb="About" label="_About" pixtype="stock" pixname="gtk-about"/>
    <menuitem name="Item 2" verb="Preferences" label="_Preferences" pixtype="stock" pixname="gtk-preferences"/>
    <menuitem name="Item 3" verb="Quit" label="_Quit" pixtype="stock" pixname="gtk-quit"/>
    </popup>
    """

    self.applet.setup_menu(self.menuxml, [ ("About",self.about_box), ("Preferences",self.prefs), ("Quit", self.cleanup) ], None)


    self.AccessO = RfkillAccess(self, self.config_ignore)

    self.update_all()
  
    # support /sys files switches
    for rf,filename in self.config_files.iteritems():
      onval = 1
      if (rf in self.onvalue):
        onval = self.onvalue[rf]
      
      try:
        foo = open(filename)
      except:
        print "ignoring sys switch for " + filename
        print "file does not exist!"
        continue

      foo.close
      tmp = SysSwitch(self, rf, filename, onval)
      self.sys_kills.append(tmp)

    applet.show_all()
    # self.load_prefs()
  
  def update_all(self):
    self.rfkills_hard = []
    self.rfkills_soft = []
    self.rfkills_name = []
    self.rfkills_idx = []
    self.rfkills_showname = []
    self.hardswitchedoff = False
    for idx, name in self.AccessO.get_rfkillall().iteritems():
      hard, soft = self.AccessO.get_state(idx)
      if hard:
        self.hardswitchedoff = True
      self.rfkills_hard.append(hard)
      self.rfkills_soft.append(soft)
      self.rfkills_name.append(name)
      self.rfkills_idx.append(idx)
      if (name in self.config_names):
        self.rfkills_showname.append(self.config_names[name])
      else:
        self.rfkills_showname.append(name)
    self.set_main_icon()
    self.update_tooltip()
    
  def update_tooltip(self):
    if (self.hardswitchedoff):
      self.ebmain.set_tooltip_text( "The devices are switched off by hardware. You have to switch the hardware switch first on!")
    else:
      self.ebmain.set_tooltip_text("Click for configuration of active devices")

  def about_box(self, event, data=None):
    authors = ["Norbert Preining <preining at logic.at>"]
    about = gtk.AboutDialog()
    about.set_name("Rfkill Applet")
    about.set_version(version)
    about.set_copyright("(C) 2009, 2010, 2011 Norbert Preining")
    about.set_authors(authors)
    #about.set_website("nothing here for now")
    #about.set_website_label("nothing here for now")
    about.run()
    about.destroy()


  def prefs(self, event, data=None):
    prefdiag = gtk.MessageDialog(buttons=gtk.BUTTONS_OK)
    prefdiag.set_property('text', "Not implemented yet!")
    prefdiag.run()
    prefdiag.destroy()

  def read_config(self, file):
    try:
      cfg = open(file, 'r')
      cfgdata = cfg.readlines()
      cfg.close()

    except:
      return

    for line in cfgdata:
      if (line.lstrip().startswith('#')):
        continue
      if line.strip() == '':
        continue
      key, val = line.strip().split('=',1)
      if val != '':
        rf,prop = key.split('.',1)
        if prop == 'onvalue':
          self.onvalue[rf] = int(val)
        elif prop == 'offvalue':
          self.offvalue[rf] = int(val)
        elif prop == 'ignore':
          self.config_ignore[rf] = val
        elif prop == 'name':
          self.config_names[rf] = val
        elif prop == 'file':
          self.config_files[rf] = val
        elif prop == 'type':
          self.config_types[rf] = val
        else:
          print "Unknown key in config file: " + line


  def set_main_icon(self):
    if (self.hardswitchedoff):
      self.icon.set_from_file(self.imagehardoff)
    else:
      self.icon.set_from_file(self.image)


  def click_menu(self, widget, event):
    if event.button == 1:
      popmenu = gtk.Menu()
      self.update_all()
      if (not(self.hardswitchedoff)):
        for idx,showname in enumerate(self.rfkills_showname):
          menu_item = gtk.CheckMenuItem(label=showname)
          menu_item.set_active(not(self.rfkills_soft[idx]))
          menu_item.show()
          menu_item.connect("toggled", self.toggle_rfkill, idx)
          popmenu.append(menu_item)
      for idx,syskill in enumerate(self.sys_kills):
        menu_item = gtk.CheckMenuItem(label=syskill.name)
        menu_item.set_active(int(syskill.state))
        menu_item.show()
        menu_item.connect("toggled", syskill.toggle_softstate)
        popmenu.append(menu_item)

      popmenu.show()
      popmenu.popup(None, None, None, event.button, event.time)


  def set_hard_switch (self, newstate):
    self.hardswitchedoff = newstate
    self.set_main_icon()
    self.update_tooltip()

  def toggle_rfkill (self, widget, idx):
    self.AccessO.toggle_softstate(self.rfkills_idx[idx])
  
  def cleanup(self, a, b):
    gtk.main_quit()
    sys.exit()


def rfkill_factory(applet, iid):
  Rfkill(applet, iid)
  return True

if len(sys.argv) == 2 and sys.argv[1] == '-d':   
  main_window = gtk.Window(gtk.WINDOW_TOPLEVEL)
  main_window.set_title("Rfkill Applet")
  main_window.connect("destroy", gtk.main_quit) 
  app = gnomeapplet.Applet()
  rfkill_factory(app, None)
  app.reparent(main_window)
  main_window.show_all()
  gtk.main()
  sys.exit()

if __name__ == '__main__':
  print('Starting factory')
  gnomeapplet.bonobo_factory("OAFIID:RfkillApplet_Factory", 
                             gnomeapplet.Applet.__gtype__, 
                             "RFKill Switch Applet", "0.1", 
                             rfkill_factory)


# vim:set tabstop=2 expandtab: #
